### PR TITLE
ceph: allow xlock state to be LOCK_PREXLOCK when putting it

### PR DESCRIPTION
**Pull request by lxbsz**
When the journal logs are successfully flushed the lockers' state could be LOCK_PREXLOCK if the inflight OP need to gather issued caps.

Fixes: https://tracker.ceph.com/issues/44565

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mds/SimpleLock.h [hunk 1]
```diff
@@ -417,6 +417,7 @@ class SimpleLock {
     ceph_assert(state == LOCK_XLOCK || state == LOCK_XLOCKDONE ||
 	   state == LOCK_XLOCKSNAP || state == LOCK_LOCK_XLOCK ||
 	   state == LOCK_LOCK  || /* if we are a leader of a peer */
+	   state == LOCK_PREXLOCK ||
```
Comment (kotreshhr) [inline src/mds/SimpleLock.h]: @lxbsz From your comment on the tracker https://tracker.ceph.com/issues/44565#note-18, How does the lock state moved to LOCK_PREXLOCK on first decrement of x from 2. Logically it make sense to move to  LOCK_PREXLOCK on x = 0 ? I don't understand the lock state machine. Sorry if it's a naive question.

Comment (lxbsz) [reply src/mds/SimpleLock.h]: No, it doesn't much matter with the `x` and the state still could be moved even it's not `0`.

This issue happened when one request tried to acquire the xlock for `ilink` locker:

Before moving the state to `LOCK_PREXLOCK` the state was `LOCK_LOCK_XLOCK`, which is a intermediate state before successfully setting it to `LOCK_XLOCK`. MDS will set the locker to a intermediate state if it couldn't move to the next target state directly, in this case it needed to revoke some caps from clients first.

From `mds/lock.cc` file we can see it's next state is `LOCK_PREXLOCK`:

```
[LOCK_LOCK_XLOCK]= { LOCK_PREXLOCK,false,LOCK_LOCK,0,   XCL, 0,   0,   0,   0,   XCL, 0,0,0,0 },
```
And just after the caps was successfully revoked and when continued moving the state to `LOCK_XLOCK`, the `eval_gather()` just set it to `LOCK_PREXLOCK`. And then later when the corresponding request was retried it should move the state to `LOCK_XLOCK`.

Comment (kotreshhr) [reply src/mds/SimpleLock.h]: Thank @lxbsz for the explaination.

**Comment by vshankar** [comment]
@lxbsz I see your testing label in this PR. Are you still testing your changes before seeking further reviews?

**Comment by lxbsz** [comment]
> @lxbsz I see your testing label in this PR. Are you still testing your changes before seeking further reviews?

@vshankar 

My tests have done and forgot to remove the label. It's ready to review. Thanks.

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2023-08-09_05:46:29-fs-wip-vshankar-testing-20230809.035933-testing-default-smithi/

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
Test update: Tests passed, preparing run wiki - will merge soon.

**Pull request merged**